### PR TITLE
rsh: Insert TERM env variable into the command to be run

### PR DESCRIPTION
**Pull request by jwendell**
Fixes the CLI part of #7424.

### CODE REVIEW

File: pkg/cmd/cli/cmd/rsh.go [hunk 1]
```diff
@@ -143,5 +144,15 @@ func (o *RshOptions) Validate() error {
...
+	// Insert the TERM into the command to be run
+	originalCommand := o.Command
+	o.Command = []string{"/bin/sh", "-c", fmt.Sprintf("TERM=vt100 %s", strings.Join(o.Command, " "))}
+
+	if o.ExecOptions.Run() == nil {
+		return nil
+	}
+
+	// If it failed, try again the original command, without the TERM addition
+	o.Command = originalCommand
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/rsh.go]: This seems super sketchy. Any rsh session returning a failure is run twice? It also assumes all commands are ok to run twice?

Comment (jwendell) [inline pkg/cmd/cli/cmd/rsh.go]: Indeed, this is a bad fix... I'll think on another approach

File: pkg/cmd/cli/cmd/rsh.go [hunk 2]
```diff
@@ -143,5 +144,15 @@ func (o *RshOptions) Validate() error {
 
 // Run starts a remote shell session on the server
 func (o *RshOptions) Run() error {
+	// Insert the TERM into the command to be run
+	originalCommand := o.Command
+	o.Command = []string{"/bin/sh", "-c", fmt.Sprintf("TERM=vt100 %s", strings.Join(o.Command, " "))}
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/rsh.go]: also, we can't assume that joining args with spaces is correct... this doesn't handle quoting/escaping

**Comment by jwendell** [comment]
Now with one-line patch. Borrowed the idea from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/exec.go#L48

**Comment by jwendell** [comment]
@liggitt still didn't like the approach?

**Comment by liggitt** [comment]
it's a lot better. @ncdc or @rhcarvalho want to weigh in?

**Comment by ncdc** [comment]
Why not use the current shell's $TERM instead of hardcoding to vt100?

**Comment by ncdc** [comment]
Also please note that this isn't 100% useful until my PR that adds support for terminal resizing lands

**Comment by ncdc** [comment]
Reviewers encouraged 😄 https://github.com/kubernetes/kubernetes/pull/25273

**Comment by jwendell** [comment]
@ncdc Now it's defaulting to current TERM's variable ;)

**Comment by rhcarvalho** [comment]
@ncdc while having resizing support will be awesome, I think this will be useful even before that!
For example, the issue description (https://github.com/openshift/origin/issues/7424) depicts the current situation when running `top` in a container:

```console
$ oc rsh docker-registry-1-x5d7z
bash-4.2$ top
TERM environment variable not set.
```

File: pkg/cmd/cli/cmd/rsh.go [hunk 3]
```diff
@@ -143,5 +144,10 @@ func (o *RshOptions) Validate() error {
 
 // Run starts a remote shell session on the server
 func (o *RshOptions) Run() error {
+	// Insert the TERM into the command to be run
+	// TODO what if the container doesn't have `env`???
```
Comment (rhcarvalho) [inline pkg/cmd/cli/cmd/rsh.go]: I assume most usages of `oc rsh` use the default command (`/bin/sh`) -- "rsh" standing for remote shell. Minimal operating systems like Busybox, and Alpine include both `/bin/sh` and `/bin/env`.

For more control, one can always use `oc exec`.

File: pkg/cmd/cli/cmd/rsh.go [hunk 4]
```diff
@@ -143,5 +144,10 @@ func (o *RshOptions) Validate() error {
...
+	term := fmt.Sprintf("TERM=%s", util.Env("TERM", "vt100"))
```
Comment (rhcarvalho) [inline pkg/cmd/cli/cmd/rsh.go]: If there is to be any default, I'd go with the same value that Docker uses for `docker run`: xterm.
[docker/docker/container/container_unix.go#L61-L63](https://github.com/docker/docker/blob/cbf785c45a8cca061028ec341320bcbd655cf1fd/container/container_unix.go#L61-L63)

**Comment by rhcarvalho** [comment]
Missing:

- [ ] Tests
- [ ] Support for `oc exec -t`

**Comment by rhcarvalho** [comment]
I think the implementation is in the wrong file.
`oc rsh` has flags to force enable or force disable TTY allocation.

We should only set TERM when we have TTY=true.

It would be "easier" to change code in `Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/exec.go`, which means fixing the problem upstream...

**Comment by rhcarvalho** [comment]
I've opened this issue upstream https://github.com/kubernetes/kubernetes/issues/28280, if it gets solved upstream for `kubectl exec`, our `oc rsh` and `oc exec -t` will inherit the fix.

**Comment by jwendell** [comment]
@rhcarvalho It seems there are no tests for rsh at all... wrt oc exec, it's just a wrapper for kubectl. If we touch it, it should be made upstream, and tbh I don't know they are like it (see the PR I sent).

@ncdc Now I've updated docs to say we send a TERM variable; And changed fallback value to xterm instead of vt100.

**Comment by ncdc** [comment]
@jwendell for now we only need the fix in `oc rsh`

**Comment by jwendell** [comment]
@ncdc, ok. this patch only touches oc rsh...

File: pkg/cmd/cli/cmd/rsh.go [hunk 5]
```diff
@@ -24,9 +25,11 @@ This command will attempt to start a shell session in a pod for the specified re
 It works with pods, deployment configs, jobs, daemon sets, and replication controllers.
 Any of the aforementioned resources (apart from pods) will be resolved to a ready pod.
 It will default to the first container if none is specified, and will attempt to use
-'/bin/bash' as the default shell. You may pass an optional command after the resource name,
+'/bin/sh' as the default shell. You may pass an optional command after the resource name,
 which will be executed instead of a login shell. A TTY will be automatically allocated
-if standard input is interactive - use -t and -T to override.
+if standard input is interactive - use -t and -T to override. A TERM variable is sent
+to the environment where the shell (or command) will be executed. By default its value
+is the same as the TERM variable from the local environment; If not found, 'xterm' is used.
```
Comment (rhcarvalho) [inline pkg/cmd/cli/cmd/rsh.go]: s/If/if ?

Maybe "if not set" would be more precise?

**Comment by rhcarvalho** [comment]
The current implementation is fine by me, @jwendell  sorry I don't know whether we have tests for `oc rsh`.

**Comment by jwendell** [comment]
So, what's necessary to have this merged? Only this minor wording thing? Anything else?

**Comment by fabianofranz** [comment]
@jwendell yes, update that minor nit and we'll get this tagged.

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
[merge]

**Pull request merged**